File: src/fitQ.c
// This code fits a quadratic regression model to the
// a moving window that is passed over an array (vector) of data
// It does so by fitting an orthogonal or near-orthogonal model and then making
// simple calculations to "correct" the parameter estimates so that the end
// result is the same as a regular quadratic regression model

// Landon Sego, 2008-01-10

/* do this first to get the right options for math.h */
#include <R_ext/Arith.h>
#include <R.h>
#include <Rmath.h>

void fitQ(double *y,   // The data array
          int *nNAy,   // Indicator array corresponding to the data array:  1 is not missing, 0 is missing
          int *n,      // The length of y
          double *x1,  // The original linear predictor (uncentered), assumed to be strictly monotonic (increasing or decreasing)
          double *x1c, // The _centered_ linear predictor, assumed to be strictly monotonic (increasing or decreasing)
          double *x2c, // The _centered_ quadratic predictor
          double *mX1, // The mean of the linear predictor (before centering)
          double *mX2, // The mean of the quadratic precitor (before centering)
          int *orthog, // Indicator of whether x1 and x2 are orthogonal
          int *bwidth, // The bandwidth of the window. The length of window is (2 * *bw + 1). Must be >= 1
          int *mwin,   // The minimum number of data points allowed for the model to be fit to the window (must be <= window length)
          int *winCen, // The indexes of y that indicate the center of the windows
          int *numWin, // The number of windows to fit
          double *a,   // The intercept coefficients of the regression model (= the mean of window)
          double *b,   // The linear coefficients
          double *c,   // The quadratic coefficients
          double *d)   // The RMSE values
{
    int win_i,           // Index of the elements in *winCen
        i,               // Index for the corresponding elements of the data array
        j,               // Index the elements in the moving window contained in 'win'
        k,               // A convenience index, a function of i, j, and bw
        bw = *bwidth,    // The bandwidth of the window
        nw = 2 * bw + 1, // The length of the window contained in 'win'
        nn = *n,         // The total length of y, the data array
        indMissing,      // Bernoulli indicator of whether there are any missing data in the window
        winC[nw],        // Bernoulli array of the elements of the window that are missing (1=present, 0=missing)
        first_x1,        // Window index of first non-missing data point in the window
        last_x1,         // Window index of last non-missing data point in the window
        getfirstx1,      // A flag to aid in finding the first_x1
        ortho = *orthog, // A flag indicating whether the model is fully orthogonal
        countI;          // An integer count of the number of non-missing data points in the window

    double win[nw],      // Array containing the moving window of data
           sumX1sq = 0,  // sum(x1_j^2)
           sumX2sq = 0,  // sum(x2_j^2)
           N = nw,       // double version of the number of data points in the window
           A,            // Value of the intercept for i_th element of y
           B,            // Value of the slope for i_th element of y
           C,            // Value of the quadratic term for i_th element of y
           resid,        // Value of the residual for the i_th element of y and the j_th element of the window
           SSE,          // Sum of squared residuals
           mx1 = *mX1,   // Non-pointer version of *mX1
           mx2 = *mX2,   // Non-pointer version of *mX2

           //  The following are used as temporary variables to calculate the parameter estimates
           count,        // Number of non-missing data points in the window
           sY,           // sum(y_j)
           sX1sq,        // sum(x1_j^2)
           sX2sq,        // sum(x2_j^2)
           sX1X2,        // sum(x1_j * x2_j)
           sX1Y,         // sum(x1_j * y_j)
           sX2Y,         // sum(x2_j * y_j)
           x1new[nw],    // New linear predictor, centered to reflect the NA's in the window
           x2new[nw],    // New quadratic predictor, centered to reflect the NA's in the window
           sx1new,       // sum(x1new_j)
           sx2new,       // sum(x2new_j)
           mx1new,       // mean(x1new_j)
           mx2new,       // mean(x2new_j)
           det;          // Suppose Z were the model matrix and X is a partition of Z such that
                         // X = Z[2:3,2:3].  Then 'det' is the determinant of X'X.

    // Sanity checks
    if (bw < 1)
      error("The bandwidth must be >= 1\n");
    if (*mwin < 3)
      error("The minimum number of non-missing data points in the window must be >= 3\n (since >= 3 points are required to fit a quadratic model)\n)");
    if (*mwin > nw)
      error("The minimum number of non-missing data points in the window\n  must be <= window length (2*bandwidth + 1)\n");
    if (nn < *mwin)
      error("The length of the data vector must be >= minimum number of\n  non-missing data points in the window\n");

    // Calculate Sum of x1^2, x2^2.  These are the only quanitites in the calculations
    // that don't depend on y (or on the missing values in y)
    for (j=0; j < nw; j++) {
      sumX1sq += x1c[j] * x1c[j];
      sumX2sq += x2c[j] * x2c[j];
    }

    // Loop over the center of each window that will be grabbed from y
    for (win_i = 0; win_i < *numWin; win_i++) {

      // Identify the index for the center of this window
      i = winCen[win_i];

      // Create the win array that contains the window of interest
      // If not near the edge of the series
      if (!(i - bw < 0 || i + bw >= nn)) {
        for (j = 0; j < nw; j++) {
          k = i - bw + j;
          win[j] = y[k];
          winC[j] = nNAy[k];
        }
      }

      // else near the edge of the series:  pad the edges with NA's as necessary
      else {
        for (j = 0; j < nw; j++) {
          k = i - bw + j;
          if (!(k < 0 || k >= nn)) {
            win[j] = y[k];
            winC[j] = nNAy[k];
          }
          else {
            win[j] = NA_REAL;
            winC[j] = 0;
          }
        }
      }

      // Initialize the variable which indicates if any of the
      // data in the window are missing
      indMissing = 0;

      // Initialize the needed variables
      sY = 0; sX1Y = 0; sX2Y = 0; SSE = 0;

      // If ortho
      if (ortho) {

        // Loop over the window to calculate sum(Y_i), sum(X1_i * Y_i), sum(X2_i * Y_i)
        // Which are the components of the least squares estimates when the predictors are orthogonal
        for (j = 0; j < nw; j++) {
          if (winC[j]) {
            sY += win[j];
            sX1Y += x1c[j] * win[j];
            sX2Y += x2c[j] * win[j];
          }
          else {
            indMissing = 1;
            break;
          }

        } // for (j = 0; ... Finish looping over the window

      } // if x1 and x2 orthogonal

      // else x1 and x2 not orthogonal
      else {

        sX1X2 = 0; det = 0;

        // Loop through the window again calculating x2new and summary statistics used
        // to calculate the coefficients
        for (j = 0; j < nw; j++) {
          if (winC[j]) {
            sY += win[j];
            sX1X2 += x1c[j] * x2c[j];
            sX1Y += x1c[j] * win[j];
            sX2Y += x2c[j] * win[j];
          }
          else {
            indMissing = 1;
            break;
          }
        }

      } // else not ortho

      // Assign the values of the coefficients if there were no missing points
      if (!indMissing) {

        // Calculate parameter estimates
        if (ortho) {
          // Parameter values needed for residuals estimates
          A = sY / N;
          B = sX1Y / sumX1sq;
          C = sX2Y / sumX2sq;
        }
        else {
          // Parameter values needed for residuals estimates
          A = sY / N;
          det = sumX1sq * sumX2sq - sX1X2 * sX1X2;
          B = (sumX2sq * sX1Y - sX1X2 * sX2Y) / det;
          C = (sumX1sq * sX2Y - sX1X2 * sX1Y) / det;
        }

        // Assign to output arrays, 'correcting' the orthogonal parameter estimates
        // so that they end up being the same as the estimates for the
        // model y = a + b*x1 + c*x1^2 + e where x1 is not centered
        a[win_i] = A - B * mx1 + C * (mx1 * mx1 - mx2);
        b[win_i] = B - 2 * C * mx1;
        c[win_i] = C;

        // Calculate sqrt(MSE) = RMSE
        if (nw > 3) {

          for (j = 0; j < nw; j++) {
            resid = win[j] - A - B * x1c[j] - C * x2c[j];
            SSE += resid * resid;
          }

          d[win_i] = sqrt(SSE / (N - 3));

        } // If nw > 3

        // when there are only 3 data points, the residual is 0.
        else if (nw == 3)
          d[win_i] = 0;
        else
          error("1: There were fewer than 3 data points in the window. This error should not have happened.\n");

      } // if there were no missing values in the window

      // If there are missing values in window re-calculate the solutions to the
      // normal equations
      else {

        // Initialize the needed variables
        countI = 0; sY = 0; sX1sq = 0; sX2sq = 0; sX1X2 = 0; sX1Y = 0; sX2Y = 0;
        sx1new = 0; sx2new = 0; first_x1 = bw + 1; last_x1 = bw - 1; getfirstx1 = 1;

        // Create the array that indicates the location of missing values,
        // count the number of values present in the window,
        // sum the x1 values in the window.  Will assume that x1 is ordered smallest to largest
        // Identify the min and max x1 values
        for (j = 0; j < nw; j++) {
          if (winC[j]) {
            sx1new += x1[j];
            countI += 1;
            last_x1 = j;
            if (getfirstx1) {
              first_x1 = j;
              getfirstx1 = 0;
            }
          }
        }

        // Get the double version of countI
        count = countI;

        // Verify that window is of sufficient size and that there is at least one data point on each side
        // of the window, where the center point qualifies as a 'side'. It ensures there won't be a window that
        // looks like this:  NA NA NA NA NA NA NA NA NA 7 2 3 l 4 7

        // The following 3 conditions are being checked:
        // 1. Num of nonNA is at least as big as the min window size
        // 2. First nonNA occurs on or before center point of window
        // 3. Last nonNA occurs on or after center point of window

        if ((countI >= *mwin) & (first_x1 <= bw) & (last_x1 >= bw)) {

          // Calculate the mean of the new x1's
          mx1new = sx1new / count;

          // Re-center the linear and quadratic predictors.  They will be orthogonal to the intercept,
          // but not to each other

          // Center the x1new and begin creating x2new
          for (j = 0; j < nw; j++) {
            if (winC[j]) {
              x1new[j] = x1[j] - mx1new;
              x2new[j] = x1new[j] * x1new[j];
              sx2new += x2new[j];
            }
          }

          // Calculate the mean of the new x2's
          mx2new = sx2new / count;

          // Loop through the window again calculating x2new and summary statistics used
          // to calculate the coefficients
          for (j = 0; j < nw; j++) {
            if (winC[j]) {
              x2new[j] = x2new[j] - mx2new;
              sY += win[j];
              sX1sq += x1new[j] * x1new[j];
              sX2sq += x2new[j] * x2new[j];
              sX1X2 += x1new[j] * x2new[j];
              sX1Y += x1new[j] * win[j];
              sX2Y += x2new[j] * win[j];
            }
          }

          // Now calculate the parameter estimates (coefficients)
          // A is orthogonal to B & C, but B & C are not orthogonal to each other
          A = sY / count;
          det = sX1sq * sX2sq - sX1X2 * sX1X2;
          B = (sX2sq * sX1Y - sX1X2 * sX2Y) / det;
          C = (sX1sq * sX2Y - sX1X2 * sX1Y) / det;

          // Assign to output arrays, 'correcting' the orthogonal parameter estimates
          // so that they end up being the same as the estimates for the
          // model y = a + b*x1 + c*x1^2 + e where x1 is not centered
          a[win_i] = A - B * mx1new + C * (mx1new * mx1new - mx2new);
          b[win_i] = B - 2 * C * mx1new;
          c[win_i] = C;

          // Calculate sqrt(MSE) = RMSE
          if (countI > 3) {

            for (j = 0; j < nw; j++) {
              if (winC[j]) {
                resid = win[j] - A - B * x1new[j] - C * x2new[j];
                SSE += resid * resid;
              }
            }
            d[win_i] = sqrt(SSE / (count - 3));

          } // If count > 3

          // when there are only 3 data points, the residual is 0.
          else if (countI == 3)
            d[win_i] = 0;
          else
            error("2: There were fewer than 3 data points in the window. This error should not have happened.\n");

        } // if the window did have sufficient data points to fit the quadratic model

        // else the window did not have sufficient data points to fit the quadratic model
        else {
          a[win_i] = NA_REAL;
          b[win_i] = NA_REAL;
          c[win_i] = NA_REAL;
          d[win_i] = NA_REAL;
        }

      } // else if (indMissing) : if there were missing values in the window

    } // for (win_i -- (initial loop)

} // end fitQ

File: R/fitQ.R
# This function calls the C quadratic fitting routine
# Fits the quadratic regression model to a moving window of
# a data vector.

# Landon Sego, January 30, 2008

# Added the 'linear.only' argument, Landon Sego, Dec 19, 2009

##' Fits a moving window quadratic (or simple linear) regression model
##' 
##' Fits a moving quadratic (or simple linear) regression model over a series
##' using a 2-sided window.
##'
##' This function dynamically accounts for the incomplete windows
##' which are caused by missing values and which occur at the beginning and end
##' of the series.
##' A quadratic regression model is used if \code{linear.only = FALSE}:
##' 
##' \code{y = a + b*x1 + c*x1^2 + e}
##' 
##' and a simple linear regression model is used if \code{linear.only = TRUE}:
##' 
##' \code{y = a + b*x1 + e}
##'
##' where \code{a}, \code{b}, and \code{c} are regression coefficients, and \code{e}
##' represents the residual error.  The regression model is fit repeatedly over a two-sided
##' moving window across the values of \code{y}. The middle element of \code{x1} is aligned with
##' an element of \code{y} and then the regression model coefficients are calculated for that element
##' of \code{y}.
##' Consequently, the size of the moving window is determined by
##' the the number of elements in \code{x1}. For example, if \code{x1 = -3:3}, the size
##' of the moving window is 7. 
##'
##' For the quadratic model, \code{fitQ} rapidly calculates the least squares estimates by centering the
##' predictors so that they are orthogonal (or nearly orthogonal, if data in
##' the window are missing).  Then the orthogonal coefficients are
##' 'backtransformed' to the original, non-orthogonal parameterization.
##' Centering and orthogonality are not required to quickly fit the simple
##' linear regression model.
##' 
##' The handling of the moving windows and missing values in this function is
##' very similar to \code{\link[Smisc:smartFilter]{Smisc::smartFilter}}.  
##'
##' Instead of a numeric vector, the \code{x1} argument can be a
##' \code{valid_fitQ_args} object (returned by \code{\link{check_fitQ_args}}), in which
##' case all the subsequent arguments to \code{fitQ} are ignored
##' (because the \code{valid_fitQ_args} object contains all those arguments). This is useful
##' because \code{fitQ} is called repeatedly by \code{\link{getFeatures}} over the same set of
##' argument values.
##' 
##' @export
##'
##' @param y A numeric vector
##' 
##' @inheritParams check_fitQ_args
##' 
##' @return An object of class \code{fitQ}, which is list with
##' the 4 vectors that contain the results of the
##' quadratic model fits:
##' \item{a}{The estimated intercepts}
##' \item{b}{The estimated linear coefficients}
##' \item{c}{The estimated quadratic coefficients. These are \code{NA} is
##'          \code{linear.only = TRUE}}
##' \item{d}{The root mean squared error (RMSE)}
##' The first element of the vectors \code{a}, \code{b}, \code{c}, and \code{d} contains the model
##' coefficients corresponding to the first data point in \code{y}.  The second element in
##' the vectors \code{a}, \code{b}, \code{c}, and \code{d} contains the model
##' coefficients corresponding to the second data point in \code{y}, etc.
##' 
##' @author Landon Sego
##' 
##' @examples
##'# Calculate the rolling window quadratic fit
##'z <- fitQ(rnorm(25), -3:3)
##'z
##'summary(z)
##'
##'# Or we can request our own summary stats
##'summary(z, stats = c("count", "mean", "kurt"))
##'

## Text and examples for the Appendix of the vignette

## # Now let's illustrate in greater detail how \code{fitQ} behaves:
## y <- rnorm(10)
## x1 <- -3:3
## x2 <- x1^2
## f <- fitQ(y, x1)
## f

## # A little function to extract out the same pieces from the \code{lm} object that
## # we wish to compare to \code{f}, the \code{fitQ} object
## extract.lm <- function(lmObject) {
    
##   x <- c(coef(lmObject), summary(lmObject)$sigma)
##   names(x) <- letters[1:4]

##   return(x)
  
## }

## # A little function for extracting the i'th element of each of the vectors in
## # a \code{fitQ} object, to compare to the output of \code{extract.lm()}:
## extract.fitQ <- function(fitQobject, i) {

##    return(unlist(lapply(fitQobject, function(x) x[i])))
    
## }

## # For the first element in \code{y}, the regression model that would be fit would be
## extract.lm(lm(y[1:4] ~ x1[4:7] + x2[4:7]))

## # Note how the middle point of \code{x1}, which, in this case, is \code{x[4] = 0},
## # is aligned with \code{y[1]}.
## # However, the first element in the vectors of \code{f} are \code{NA}:
## extract.fitQ(f, 1)

## # This is because the default value of \code{min.window = 5} requires
## # there be at least 5 non-missing data points to fit the regression model, whereas there were only 4 in
## # this case.  For the second element of \code{y}, the regression model is constructed
## # by aligning the middle point of \code{x1} with the second element of \code{y}: 
## extract.lm(lm(y[1:5] ~ x1[3:7] + x2[3:7]))

## # And the second vector elements from \code{fitQ()} match the results from the \code{lm()} fit:
## extract.fitQ(f, 2)

## # Here's what we see for the third element in \code{y}:
## extract.lm(lm(y[1:6] ~ x1[2:7] + x2[2:7]))
## extract.fitQ(f, 3)

## # Once we get to the fourth element, we are able to fit a model with the whole window
## # of seven elements (because \code{x1} has seven elements):
## extract.lm(lm(y[1:7] ~ x1 + x2))
## extract.fitQ(f, 4)

## # And all subsequent elements in the body of \code{y} use the whole window as well,
## extract.lm(lm(y[2:8] ~ x1 + x2))
## extract.fitQ(f, 5)

## # Until we get to the end of \code{y}, in which case the the last elements of \code{x1} start
## # dropping off.  Here's the fit for the 8th element of \code{y}:
## extract.lm(lm(y[5:10] ~ x1[1:6] + x2[1:6]))
## extract.fitQ(f, 8)

## # And the 9th element of \code{y}:
## extract.lm(lm(y[6:10] ~ x1[1:5] + x2[1:5]))
## extract.fitQ(f, 9)

## # An illustration how how \code{fitQ()} handles \code{NA}'s is also in order.  Let's insert an
## # \code{NA} into \code{y}:
## y[3] <- NA
## f <- fitQ(y, x1)
## f

## # The first elements of the \code{fitQ} object are \code{NA} because there are only 3 non-missing
## # data points in the window for the first element:
## sum(!is.na(y[1:4]))

## # The second elements
## # are \code{NA} because there are only 4 non-missing data points in the window that is centered on the
## # second element of y:
## sum(!is.na(y[1:5]))

## # However, for the third element of y, we now have 5 non-missing data points, which is the default
## # value for \code{min.window}, the minimum allowable number of non-missing values:
## sum(!is.na(y[1:6]))

## # And we see that the models agree
## extract.lm(lm(y[1:6] ~ x1[2:7] + x2[2:7]))
## extract.fitQ(f, 3)

fitQ <- function(y,
                 x1 = -10:10,
                 min.window = 5,
                 start = 1,
                 skip = 1,
                 linear.only = FALSE) {
    
  # Check whether x1 is a 'valid_fitQ_args' object.  If not, check the args
  if (inherits(x1, "valid_fitQ_args")) {
    a <- x1
  }
  else {
    a <- check_fitQ_args(x1 = x1, min.window = min.window, start = start,
                         skip = skip, linear.only = linear.only)
  }

  # Checks on y that need to always be done
  stopifnot(is.numeric(y),
            is.vector(y),
            length(y) > 3)
  
  # Number of elements in y
  n <- length(y)

  # Checks that depend on y
  if (a$start > n) {
    stop("'start' must be <= 'length(y)'")
  }

  if (n < a$min.window) {
    stop("'length(y)' must be >= 'min.window'")
  }
  
  # Index of window centers
  win.centers <- seq(a$start, n, by = a$skip)
  num.windows <- length(win.centers)

  # Quadratic fit
  if (!a$linear.only) {

    # Center the x1
    mx1 <- mean(a$x1)
    x1.c <- a$x1 - mx1
    
    # Create quadratic predictor variable and center it
    x2 <- x1.c ^ 2
    mx2 <- mean(x2)
    x2.c <- x2 - mx2
  
    # Determine whether x1 and x2 are orthogonal
    orthog <- round(t(x1.c) %*% x2.c, 10) == 0
  
    # Call the C 'fitQ' method that fits the quadratic model
    # over the entire data vector and return a list of the a's, b's, c's, and d's.
    out <- .C("fitQ",
              as.double(y),
              as.integer(!is.na(y)),
              as.integer(n),
              as.double(a$x1),
              as.double(x1.c),
              as.double(x2.c),
              as.double(mx1),
              as.double(mx2),
              as.integer(orthog),
              as.integer(a$bw),
              as.integer(a$min.window),
              as.integer(win.centers - 1),
              as.integer(num.windows),
              a = double(num.windows),
              b = double(num.windows),
              c = double(num.windows),
              d = double(num.windows),
              NAOK = TRUE)[letters[1:4]]

  }

  # Linear fit
  else {
    
    out1 <- .C("fitL",
               as.double(y),
               as.integer(!is.na(y)),
               as.integer(n),
               as.double(a$x1),
               as.double(mean(a$x1)),
               as.double(t(a$x1) %*% a$x1),
               as.integer(a$bw),
               as.integer(a$min.window),
               as.integer(win.centers - 1),
               as.integer(num.windows),
               a = double(num.windows),
               b = double(num.windows),
               d = double(num.windows),
               NAOK = TRUE)[c("a", "b", "d")]

    out <- list(a = out1$a, b = out1$b, c = rep(NA, num.windows), d = out1$d)
 
  }

  # Give it a class
  class(out) <- c("fitQ", class(out))
  
  return(out)
  
} # fitQ

##' @method summary fitQ
##'
##' @describeIn fitQ Calculates summary statistics for a \code{fitQ} object, returning a
##' named vector with summary statistics.  The names take the form [coefficient].[stat],
##' where the coefficient is one of "a", "b", "c", or "d", and the stat is the statistic
##' required in the \code{stats} argument of the \code{summary} method.
##'
##' @param object An object of class \code{fit} returned by \code{\link{fitQ}}.
##' 
##' @param \dots Additional (unused) arguments required for consistency of S3 methods
##' 
##' @param stats A character vector of summary statistics that are valid for
##' \code{\link{summaryStats}}. Alternatively, the object returned by
##' \code{\link{summaryStats}} may also be supplied for this argument.
##'
##' @export

# Summary method for fitQ
summary.fitQ <- function(object, ...,
                         stats = c("min", "q1", "mean", "med", "q3",
                                   "max", "sd", "count")) {

  # Create the summary stat function, unless it has been passed in
  if (!inherits(stats, "summaryStats_function")) {
    stats <- summaryStats(stats)
  }

  # Calculate the summary statistics
  out <- unlist(lapply(object, stats))

  return(out)

} # summary.fitQ
